File: src/main/java/io/tolgee/dtos/PathDTO.java
package io.tolgee.dtos;

import io.tolgee.helpers.TextHelper;

import java.util.Collections;
import java.util.LinkedList;
import java.util.List;
import java.util.stream.Collectors;

public class PathDTO {
    public static final char DELIMITER = '.';

    private LinkedList<String> fullPath = new LinkedList<>();

    private PathDTO() {
    }

    public static PathDTO fromFullPath(String fullPath) {
        PathDTO pathDTO = new PathDTO();
        pathDTO.add(TextHelper.splitOnNonEscapedDelimiter(fullPath, DELIMITER));
        return pathDTO;
    }

    public static PathDTO fromFullPath(List<String> path) {
        PathDTO pathDTO = new PathDTO();
        pathDTO.add(path);
        return pathDTO;
    }

    public static PathDTO fromPathAndName(String path, String name) {
        PathDTO pathDTO = new PathDTO();
        List<String> items = TextHelper.splitOnNonEscapedDelimiter(path, DELIMITER);
        if (path.isEmpty()) {
            items = Collections.emptyList();
        }
        pathDTO.add(items);
        pathDTO.add(name);
        return pathDTO;
    }

    public static PathDTO fromPathAndName(List<String> path, String name) {
        PathDTO pathDTO = new PathDTO();
        pathDTO.add(path);
        pathDTO.add(name);
        return pathDTO;
    }

    public String getName() {
        return fullPath.getLast();
    }

    public String getFullPathString() {
        return getFullPath().stream().map(i -> i.replaceAll("\\" + DELIMITER, "\\\\" + DELIMITER)).collect(Collectors.joining("."));
    }

    public List<String> getPath() {
        LinkedList<String> path = new LinkedList<>(this.fullPath);
        path.removeLast();
        return path;
    }

    public List<String> getFullPath() {
        return new LinkedList<>(this.fullPath);
    }

    private void add(String item) {
        this.fullPath.add(item);
    }

    private void add(List<String> list) {
        this.fullPath.addAll(list);
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;

        PathDTO pathDTO = (PathDTO) o;

        return getFullPath().equals(pathDTO.getFullPath());
    }

    @Override
    public int hashCode() {
        return getFullPath().hashCode();
    }

    public void setFullPath(LinkedList<String> fullPath) {
        this.fullPath = fullPath;
    }

    public String toString() {
        return "PathDTO(fullPath=" + this.getFullPath() + ")";
    }
}

File: backend/core/src/main/kotlin/io/tolgee/dtos/PathDTO.kt
package io.tolgee.dtos

import io.tolgee.helpers.TextHelper.splitOnNonEscapedDelimiter
import java.util.LinkedList
import java.util.stream.Collectors

class PathDTO private constructor() {
  var fullPath: MutableList<String> = mutableListOf()

  val name: String
    get() = fullPath.last()
  val fullPathString: String
    get() =
      fullPath
        .stream()
        .map { i: String ->
          i.replace(
            ("\\" + DELIMITER).toRegex(),
            "\\\\" + DELIMITER,
          )
        }.collect(
          Collectors.joining("."),
        )
  val path: List<String>
    get() {
      val path = LinkedList(fullPath)
      path.removeLast()
      return path
    }

  private fun add(item: String) {
    fullPath.add(item)
  }

  private fun add(list: List<String>) {
    fullPath.addAll(list)
  }

  fun setFullPath(fullPath: LinkedList<String>) {
    this.fullPath = fullPath
  }

  override fun toString(): String {
    return "PathDTO(fullPath=" + fullPath + ")"
  }

  companion object {
    const val DELIMITER = '.'

    fun fromFullPath(fullPath: String?): PathDTO {
      val pathDTO = PathDTO()
      pathDTO.add(splitOnNonEscapedDelimiter(fullPath!!, DELIMITER))
      return pathDTO
    }

    fun fromFullPath(path: List<String>): PathDTO {
      val pathDTO = PathDTO()
      pathDTO.add(path)
      return pathDTO
    }

    fun fromPathAndName(
      path: String,
      name: String,
    ): PathDTO {
      val pathDTO = PathDTO()
      var items = splitOnNonEscapedDelimiter(path, DELIMITER)
      if (path.isEmpty()) {
        items = emptyList()
      }
      pathDTO.add(items)
      pathDTO.add(name)
      return pathDTO
    }

    fun fromPathAndName(
      path: List<String>,
      name: String,
    ): PathDTO {
      val pathDTO = PathDTO()
      pathDTO.add(path)
      pathDTO.add(name)
      return pathDTO
    }
  }
}
